File: frontend/src/lib/services/agents.ts
// Agents Service
import { supabase } from '../supabase';
import type { Database } from '../supabase';

type UserAgent = Database['public']['Tables']['user_agents']['Row'];
type UserAgentInsert = Database['public']['Tables']['user_agents']['Insert'];
type UserAgentUpdate = Database['public']['Tables']['user_agents']['Update'];

export const agentsService = {
    // Get all agents for user
    async getAgents(userId: string) {
        const { data, error } = await supabase
            .from('user_agents')
            .select('*')
            .eq('user_id', userId)
            .order('created_at', { ascending: false });

        if (error) throw error;
        return data || [];
    },

    // Get agent by ID
    async getAgentById(id: string) {
        const { data, error } = await supabase
            .from('user_agents')
            .select('*')
            .eq('id', id)
            .single();

        if (error) throw error;
        return data;
    },

    // Create agent
    async createAgent(agent: Omit<UserAgentInsert, 'id'>) {
        const { data, error } = await supabase
            .from('user_agents')
            .insert(agent)
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    // Update agent
    async updateAgent(id: string, updates: UserAgentUpdate) {
        const { data, error } = await supabase
            .from('user_agents')
            .update(updates)
            .eq('id', id)
            .select()
            .single();

        if (error) throw error;
        return data;
    },

    // Delete agent
    async deleteAgent(id: string) {
        const { error } = await supabase
            .from('user_agents')
            .delete()
            .eq('id', id);

        if (error) throw error;
    },
};

File: backend/app/routers/agents.py
"""Agents router — CRUD for AI agent configurations."""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.supabase_client import get_supabase
from app.middleware.auth import get_current_user, AuthUser

router = APIRouter()


class AgentCreate(BaseModel):
    name: str
    description: str = ""
    system_prompt: str = ""
    provider: str = "openrouter"
    model: str = ""
    tools: list[str] = []
    temperature: float = 0.7


class AgentUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    system_prompt: str | None = None
    provider: str | None = None
    model: str | None = None
    tools: list[str] | None = None
    temperature: float | None = None


class AgentResponse(BaseModel):
    id: str
    user_id: str
    name: str
    description: str
    system_prompt: str
    provider: str
    model: str
    tools: list[str]
    temperature: float
    created_at: str
    updated_at: str


@router.get("/", response_model=list[AgentResponse])
async def list_agents(user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    result = await sb.select("user_agents", filters={"user_id": user.id}, order="-updated_at")
    return result or []


@router.post("/", response_model=AgentResponse, status_code=201)
async def create_agent(body: AgentCreate, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    rows = await sb.insert("user_agents", {
        "user_id": user.id,
        "name": body.name,
        "description": body.description,
        "system_prompt": body.system_prompt,
        "provider": body.provider,
        "model": body.model,
        "tools": body.tools,
        "temperature": body.temperature,
    })
    if not rows:
        raise HTTPException(status_code=500, detail="Failed to create agent")
    return rows[0]


@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(agent_id: str, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    try:
        result = await sb.select("user_agents", filters={"id": agent_id, "user_id": user.id}, single=True)
    except Exception:
        raise HTTPException(status_code=404, detail="Agent not found")
    if not result:
        raise HTTPException(status_code=404, detail="Agent not found")
    return result


@router.patch("/{agent_id}", response_model=AgentResponse)
async def update_agent(agent_id: str, body: AgentUpdate, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    updates = body.model_dump(exclude_none=True)
    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")
    rows = await sb.update("user_agents", updates, {"id": agent_id, "user_id": user.id})
    if not rows:
        raise HTTPException(status_code=404, detail="Agent not found")
    return rows[0]


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(agent_id: str, user: AuthUser = Depends(get_current_user)):
    sb = get_supabase()
    await sb.delete("user_agents", {"id": agent_id, "user_id": user.id})
